curp/src/server/state.rs

```diff
@@ -117,9 +120,37 @@ impl<C: Command + 'static> State<C> {
         self.last_applied < self.commit_index
     }
 
+    /// Do cleanup and redo logs when the leader retires
+    #[allow(clippy::indexing_slicing)] // log.len() >= 1 because we have a fake log[0], indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+    fn leader_retires(&mut self) {
+        // when a leader retires, it should wipe up speculatively executed cmds by resetting and re-executing
+        self.cmd_exe_tx.send_reset();
+
+        let mut board_w = self.cmd_board.write();
+        board_w.clear();
+
+        for i in 0..=self.commit_index {
+            let log = &self.log[i];
+            for cmd in log.cmds() {
+                assert!(
+                    board_w.needs_exe.insert(cmd.id().clone()),
+                    "shouldn't insert needs_exe twice"
+                );
+                // FIXME: do we need to call after sync or that execution is enough?
+                self.cmd_exe_tx
```

Comment:
After viewing xline's code, I guess after sync is still needed because cmds like `PutRequest` will only be truely processed in after sync. See `sync_request` in `kv_store`.